subsys/debug/ppi_trace/ppi_trace.c

```diff
@@ -0,0 +1,201 @@
+ * ----------------------------------------------------------------
+ */
+#define HANDLE_FLAG (1 << 31)
+#define PAIR_FLAG (1 << 30)
+
+#define HANDLE_ENCODE(value) (void *)((u32_t)value | HANDLE_FLAG)
+#define PACK_CHANNELS(start_ch, stop_ch) (PAIR_FLAG | start_ch | (stop_ch << 8))
+
+#define IS_VALID_HANDLE(handle) ((u32_t)handle & HANDLE_FLAG)
+#define IS_PAIR(handle) ((u32_t)handle & PAIR_FLAG)
+
+#define GET_CH(handle) ((u32_t)handle & 0xFF)
+#define GET_START_CH(handle) GET_CH(handle)
+#define GET_STOP_CH(handle) (((u32_t)handle >> 8) & 0xFF)
+
+/** @brief Allocate (D)PPI channel. */
+static nrfx_err_t ppi_alloc(u8_t *ch, u32_t evt)
+{
+	nrfx_err_t err;
+#ifdef DPPI_PRESENT
+	if (*PUBLISH_ADDR(evt) != 0) {
+		*ch = *PUBLISH_ADDR(evt) & 0x0F;
+		err = NRFX_SUCCESS;
+	} else {
+		err = nrfx_dppi_channel_alloc(ch);
+	}
+#else
+	err = nrfx_ppi_channel_alloc((nrf_ppi_channel_t *)ch);
+#endif
+	return err;
+}
+
+/** @brief Set task and event on (D)PPI channel. */
+static void ppi_assign(u8_t ch, u32_t evt, u32_t task)
+{
+#ifdef DPPI_PRESENT
+	*SUBSCRIBE_ADDR(task) = 0x80000000 | (u32_t)ch;
+	*PUBLISH_ADDR(evt) = 0x80000000 | (u32_t)ch;
+#else
+	(void)nrfx_ppi_channel_assign(ch, evt, task);
+#endif
+}
+
+/** @brief Enable (D)PPI channel. */
+static void ppi_enable(u8_t ch)
+{
+#ifdef DPPI_PRESENT
+	(void)nrfx_dppi_channel_enable(ch);
+#else
+	(void)nrfx_ppi_channel_enable(ch);
+#endif
+}
+
+/** @brief Disable (D)PPI channel. */
+static void ppi_disable(u8_t ch)
+{
+#ifdef DPPI_PRESENT
+	(void)nrfx_dppi_channel_disable(ch);
+#else
+	(void)nrfx_ppi_channel_disable(ch);
+#endif
+}
+
+/** @brief Allocate GPIOTE channel.
+ *
+ * @param pin Pin.
+ *
+ * @return Allocated channel or -1 if failed to allocate.
+ */
+static int gpiote_channel_alloc(u32_t pin)
+{
+	for (u8_t channel = 0; channel < GPIOTE_CH_NUM; ++channel) {
+		if (!nrf_gpiote_te_is_enabled(channel)) {
+			nrf_gpiote_task_configure(channel, pin,
+						  NRF_GPIOTE_POLARITY_TOGGLE,
+						  NRF_GPIOTE_INITIAL_VALUE_LOW);
+			nrf_gpiote_task_enable(channel);
+			return channel;
+		}
+	}
+
+	return -1;
+}
+
+void *ppi_trace_config(u32_t pin, u32_t evt)
+{
+	int err;
+	u32_t task;
+	int gpiote_ch;
+	nrf_gpiote_tasks_t task_id = NRF_GPIOTE_TASKS_OUT_0;
+	u8_t ppi_ch;
+
+	err = ppi_alloc(&ppi_ch, evt);
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("Failed to allocate PPI channel.");
+		return NULL;
+	}
+
+	gpiote_ch = gpiote_channel_alloc(pin);
+	if (gpiote_ch < 0) {
+		LOG_ERR("Failed to allocate GPIOTE channel.");
+		return NULL;
+	}
+
+	task_id += gpiote_ch*sizeof(u32_t);
+	task = nrf_gpiote_task_addr_get(task_id);
+	ppi_assign(ppi_ch, evt, task);
+
+	return HANDLE_ENCODE(ppi_ch);
+}
+
+void *ppi_trace_pair_config(u32_t pin, u32_t start_evt, u32_t stop_evt)
+{
+	int err;
+	u32_t task;
+	u8_t start_ch;
+	u8_t stop_ch;
+	int gpiote_ch;
+	nrf_gpiote_tasks_t task_set_id = NRF_GPIOTE_TASKS_SET_0;
+	nrf_gpiote_tasks_t task_clr_id = NRF_GPIOTE_TASKS_CLR_0;
```

Comment:
This makes the whole module unavailable for nRF51 Series. I think it would be nice to clearly state it in the documentation. Alternatively, only this function could be made unavailable for '51s. Have you considered this?

Reply:
fixed. setting pair is not supported on nrf51, simple tracing is.